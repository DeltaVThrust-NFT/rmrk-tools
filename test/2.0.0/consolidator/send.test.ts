import { Consolidator } from "../../../src/rmrk2.0.0";
import {
  createCollectionMock,
  getBlockCallsMock,
  getBobKey,
  getRemarksFromBlocksMock,
  mintNftMock,
  mintNftMock2,
  mintNftMock3,
} from "../mocks";
import { cryptoWaitReady } from "@polkadot/util-crypto";

beforeAll(async () => {
  return await cryptoWaitReady();
});

describe("rmrk2.0.0 Consolidator: Send NFT to other NFT", () => {
  const getSetupRemarks = () => [
    ...getBlockCallsMock(createCollectionMock().create()),
    ...getBlockCallsMock(mintNftMock().mint()),
    ...getBlockCallsMock(mintNftMock2().mint()),
  ];

  it("Add new NFT as child of first NFT", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftMock2(4).send(mintNftMock(3).getId())),
    ]);
    const consolidator = new Consolidator();
    expect(await consolidator.consolidate(remarks)).toMatchSnapshot();
  });

  it("Add new NFT as child of first NFT and then send it to different parent NFT", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftMock3().mint()),
      ...getBlockCallsMock(mintNftMock2(4).send(mintNftMock(3).getId())),
      ...getBlockCallsMock(mintNftMock2(4).send(mintNftMock3(5).getId())),
    ]);
    const consolidator = new Consolidator();
    expect(await consolidator.consolidate(remarks)).toMatchSnapshot();
  });

  it("rootowner should be updated recursivly", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getBlockCallsMock(createCollectionMock().create()),
      ...getBlockCallsMock(mintNftMock().mint()),
      ...getBlockCallsMock(mintNftMock2().mint(mintNftMock(3).getId())),
      ...getBlockCallsMock(mintNftMock3().mint(mintNftMock2(4).getId())),
      ...getBlockCallsMock(mintNftMock(3).send(getBobKey().address)),
    ]);
    const consolidator = new Consolidator();
    expect(await consolidator.consolidate(remarks)).toMatchSnapshot();
  });

  it("Should set NFT as pending if sent from non parent owner", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftMock2(4).send(getBobKey().address)), // Send to Bob first
      ...getBlockCallsMock(
        mintNftMock2(4).send(mintNftMock(3).getId()),
        getBobKey().address
      ),
    ]);
    const consolidator = new Consolidator();
    expect(await consolidator.consolidate(remarks)).toMatchSnapshot();
  });

  it("Should invalidate SEND to non-existant NFT", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftMock2(99).send(getBobKey().address)), // Send to Bob first
    ]);
    const consolidator = new Consolidator();
    const consolidated = await consolidator.consolidate(remarks);
    expect(consolidated.invalid[0].message).toEqual(
      "[SEND] Attempting to send non-existant NFT 99-d43593c715a56da27d-KANARIABIRDS-KANR-00000888"
    );
  });

  it("Should invalidate SEND to burned NFT", async () => {
    const nftMock = mintNftMock2(4);
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(nftMock.burn()),
      ...getBlockCallsMock(nftMock.send(getBobKey().address)), // Send to Bob first
    ]);
    const consolidator = new Consolidator();
    const consolidated = await consolidator.consolidate(remarks);
    expect(consolidated.invalid[0].message).toEqual(
      "[SEND] Attempting to send burned NFT 4-d43593c715a56da27d-KANARIABIRDS-KANR-00000888"
    );
  });

  it("Should invalidate SEND of non-owned child NFT", async () => {
    const nftMock = mintNftMock2(4);
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(mintNftMock3().mint(), getBobKey().address),
      ...getBlockCallsMock(mintNftMock3(5).send(mintNftMock(3).getId())), // Send to another NFT first
      ...getBlockCallsMock(nftMock.send(getBobKey().address)), // Send to Bob
    ]);
    const consolidator = new Consolidator();
    const consolidated = await consolidator.consolidate(remarks);
    expect(consolidated.invalid[0].message).toEqual(
      "[SEND] Attempting to send non-owned NFT 5-d43593c715a56da27d-KANARIABIRDS-KANR-00000999, real owner: 5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty"
    );
  });

  it("Should invalidate SEND to non-existant NFT", async () => {
    const nftMock = mintNftMock2(4);
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(nftMock.send(mintNftMock(5).getId())),
    ]);
    const consolidator = new Consolidator();
    const consolidated = await consolidator.consolidate(remarks);
    expect(consolidated.invalid[0].message).toEqual(
      "[SEND] Attempting to send NFT to a non existing NFT 5-d43593c715a56da27d-KANARIABIRDS-KANR-00000777."
    );
  });
});
