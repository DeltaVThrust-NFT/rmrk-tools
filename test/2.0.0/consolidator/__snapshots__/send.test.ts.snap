// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rmrk2.0.0 Consolidator: Send NFT to other NFT Add new NFT as child of first NFT 1`] = `
Object {
  "bases": Array [],
  "collections": Array [
    Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
  ],
  "invalid": Array [],
  "nfts": Array [
    Object {
      "block": 3,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "metadata": undefined,
      "owner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "pending": false,
      "priority": Array [],
      "properties": Object {},
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "sn": "00000777",
      "symbol": "KANR",
      "transferable": 1,
    },
    Object {
      "block": 4,
      "burned": "",
      "changes": Array [
        Object {
          "block": 5,
          "caller": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "field": "owner",
          "new": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
          "old": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "opType": "SEND",
        },
      ],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000888",
      "metadata": undefined,
      "owner": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "pending": false,
      "priority": Array [],
      "properties": Object {},
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "sn": "00000888",
      "symbol": "KANR",
      "transferable": 1,
    },
  ],
}
`;

exports[`rmrk2.0.0 Consolidator: Send NFT to other NFT Add new NFT as child of first NFT and then send it to different parent NFT 1`] = `
Object {
  "bases": Array [],
  "collections": Array [
    Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
  ],
  "invalid": Array [],
  "nfts": Array [
    Object {
      "block": 3,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "metadata": undefined,
      "owner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "pending": false,
      "priority": Array [],
      "properties": Object {},
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "sn": "00000777",
      "symbol": "KANR",
      "transferable": 1,
    },
    Object {
      "block": 4,
      "burned": "",
      "changes": Array [
        Object {
          "block": 6,
          "caller": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "field": "owner",
          "new": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
          "old": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "opType": "SEND",
        },
        Object {
          "block": 7,
          "caller": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "field": "owner",
          "new": "5-d43593c715a56da27d-KANARIABIRDS-KANR-00000999",
          "old": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
          "opType": "SEND",
        },
      ],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000888",
      "metadata": undefined,
      "owner": "5-d43593c715a56da27d-KANARIABIRDS-KANR-00000999",
      "pending": false,
      "priority": Array [],
      "properties": Object {},
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "sn": "00000888",
      "symbol": "KANR",
      "transferable": 1,
    },
    Object {
      "block": 5,
      "burned": "",
      "changes": Array [],
      "children": Array [
        Object {
          "equipped": "",
          "id": "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000888",
          "pending": false,
        },
      ],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "5-d43593c715a56da27d-KANARIABIRDS-KANR-00000999",
      "metadata": undefined,
      "owner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "pending": false,
      "priority": Array [],
      "properties": Object {},
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "sn": "00000999",
      "symbol": "KANR",
      "transferable": 1,
    },
  ],
}
`;

exports[`rmrk2.0.0 Consolidator: Send NFT to other NFT Should set NFT as pending if sent from non parent owner 1`] = `
Object {
  "bases": Array [],
  "collections": Array [
    Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
  ],
  "invalid": Array [],
  "nfts": Array [
    Object {
      "block": 3,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "metadata": undefined,
      "owner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "pending": false,
      "priority": Array [],
      "properties": Object {},
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "sn": "00000777",
      "symbol": "KANR",
      "transferable": 1,
    },
    Object {
      "block": 4,
      "burned": "",
      "changes": Array [
        Object {
          "block": 5,
          "caller": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "field": "owner",
          "new": "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
          "old": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "opType": "SEND",
        },
        Object {
          "block": 6,
          "caller": "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
          "field": "owner",
          "new": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
          "old": "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
          "opType": "SEND",
        },
      ],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000888",
      "metadata": undefined,
      "owner": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "pending": false,
      "priority": Array [],
      "properties": Object {},
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "sn": "00000888",
      "symbol": "KANR",
      "transferable": 1,
    },
  ],
}
`;

exports[`rmrk2.0.0 Consolidator: Send NFT to other NFT Should set NFT as pending if sent to non owned NFT 1`] = `
Object {
  "bases": Array [],
  "collections": Array [
    Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
  ],
  "invalid": Array [],
  "nfts": Array [
    Object {
      "block": 3,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "metadata": undefined,
      "owner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "pending": false,
      "priority": Array [],
      "properties": Object {},
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "sn": "00000777",
      "symbol": "KANR",
      "transferable": 1,
    },
    Object {
      "block": 4,
      "burned": "",
      "changes": Array [
        Object {
          "block": 6,
          "caller": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "field": "owner",
          "new": "5-d43593c715a56da27d-KANARIABIRDS-KANR-00000999",
          "old": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "opType": "SEND",
        },
      ],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000888",
      "metadata": undefined,
      "owner": "5-d43593c715a56da27d-KANARIABIRDS-KANR-00000999",
      "pending": false,
      "priority": Array [],
      "properties": Object {},
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
      "sn": "00000888",
      "symbol": "KANR",
      "transferable": 1,
    },
    Object {
      "block": 5,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "5-d43593c715a56da27d-KANARIABIRDS-KANR-00000999",
      "metadata": undefined,
      "owner": "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
      "pending": false,
      "priority": Array [],
      "properties": Object {},
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
      "sn": "00000999",
      "symbol": "KANR",
      "transferable": 1,
    },
  ],
}
`;

exports[`rmrk2.0.0 Consolidator: Send NFT to other NFT rootowner should be updated recursivly 1`] = `
Object {
  "bases": Array [],
  "collections": Array [
    Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
  ],
  "invalid": Array [],
  "nfts": Array [
    Object {
      "block": 3,
      "burned": "",
      "changes": Array [
        Object {
          "block": 6,
          "caller": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "field": "owner",
          "new": "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
          "old": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "opType": "SEND",
        },
      ],
      "children": Array [
        Object {
          "equipped": "",
          "id": "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000888",
          "pending": false,
        },
      ],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "metadata": undefined,
      "owner": "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
      "pending": false,
      "priority": Array [],
      "properties": Object {},
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
      "sn": "00000777",
      "symbol": "KANR",
      "transferable": 1,
    },
    Object {
      "block": 4,
      "burned": "",
      "changes": Array [],
      "children": Array [
        Object {
          "equipped": "",
          "id": "5-d43593c715a56da27d-KANARIABIRDS-KANR-00000999",
          "pending": false,
        },
      ],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000888",
      "metadata": undefined,
      "owner": "3-d43593c715a56da27d-KANARIABIRDS-KANR-00000777",
      "pending": false,
      "priority": Array [],
      "properties": Object {},
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
      "sn": "00000888",
      "symbol": "KANR",
      "transferable": 1,
    },
    Object {
      "block": 5,
      "burned": "",
      "changes": Array [],
      "children": Array [],
      "collection": "d43593c715a56da27d-KANARIABIRDS",
      "forsale": 0n,
      "id": "5-d43593c715a56da27d-KANARIABIRDS-KANR-00000999",
      "metadata": undefined,
      "owner": "4-d43593c715a56da27d-KANARIABIRDS-KANR-00000888",
      "pending": false,
      "priority": Array [],
      "properties": Object {},
      "reactions": Object {},
      "resources": Array [],
      "rootowner": "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
      "sn": "00000999",
      "symbol": "KANR",
      "transferable": 1,
    },
  ],
}
`;
